File: ICO_Template/ICO_Template.cs
﻿using Neo.SmartContract.Framework;
using Neo.SmartContract.Framework.Services.Neo;
using Neo.SmartContract.Framework.Services.System;
using System;
using System.ComponentModel;
using System.Numerics;

namespace Neo.SmartContract
{
    public class ICO_Template : Framework.SmartContract
    {
        //Token Settings
        [DisplayName("name")]
        public static string Name() => "name of the token";
        [DisplayName("symbol")]
        public static string Symbol() => "SymbolOfTheToken";
        public static readonly byte[] Owner = "ATrzHaicmhRj15C3Vv6e6gLfLqhSD2PtTr".ToScriptHash();
        [DisplayName("decimals")]
        public static byte Decimals() => 8;
        private const ulong factor = 100000000; //decided by Decimals()
        private const ulong neo_decimals = 100000000;

        //ICO Settings
        private static readonly byte[] neo_asset_id = { 155, 124, 255, 218, 166, 116, 190, 174, 15, 147, 14, 190, 96, 133, 175, 144, 147, 229, 254, 86, 179, 74, 92, 34, 12, 205, 207, 110, 252, 51, 111, 197 };
        private const ulong total_amount = 100000000 * factor; // total token amount
        private const ulong pre_ico_cap = 30000000 * factor; // pre ico token amount
        private const ulong basic_rate = 1000 * factor;
        private const int ico_start_time = 1506787200;
        private const int ico_end_time = 1538323200;

        [DisplayName("transfer")]
        public static event Action<byte[], byte[], BigInteger> Transferred;

        [DisplayName("refund")]
        public static event Action<byte[], BigInteger> Refund;

        public static Object Main(string operation, params object[] args)
        {
            if (Runtime.Trigger == TriggerType.Verification)
            {
                if (Owner.Length == 20)
                {
                    // if param Owner is script hash
                    return Runtime.CheckWitness(Owner);
                }
                else if (Owner.Length == 33)
                {
                    // if param Owner is public key
                    byte[] signature = operation.AsByteArray();
                    return VerifySignature(signature, Owner);
                }
            }
            else if (Runtime.Trigger == TriggerType.Application)
            {
                if (operation == "deploy") return Deploy();
                if (operation == "mintTokens") return MintTokens();
                if (operation == "totalSupply") return TotalSupply();
                if (operation == "name") return Name();
                if (operation == "symbol") return Symbol();
                if (operation == "transfer")
                {
                    if (args.Length != 3) return false;
                    byte[] from = (byte[])args[0];
                    byte[] to = (byte[])args[1];
                    BigInteger value = (BigInteger)args[2];
                    return Transfer(from, to, value);
                }
                if (operation == "balanceOf")
                {
                    if (args.Length != 1) return 0;
                    byte[] account = (byte[])args[0];
                    return BalanceOf(account);
                }
                if (operation == "decimals") return Decimals();
            }
            //you can choice refund or not refund
            byte[] sender = GetSender();
            ulong contribute_value = GetContributeValue();
            if (contribute_value > 0 && sender.Length != 0)
            {
                Refund(sender, contribute_value);
            }
            return false;
        }

        // initialization parameters, only once
        // 初始化参数
        [DisplayName("deploy")]
        public static bool Deploy()
        {
            byte[] total_supply = Storage.Get(Storage.CurrentContext, "totalSupply");
            if (total_supply.Length != 0) return false;
            Storage.Put(Storage.CurrentContext, Owner, pre_ico_cap);
            Storage.Put(Storage.CurrentContext, "totalSupply", pre_ico_cap);
            Transferred(null, Owner, pre_ico_cap);
            return true;
        }

        // The function MintTokens is only usable by the chosen wallet
        // contract to mint a number of tokens proportional to the
        // amount of neo sent to the wallet contract. The function
        // can only be called during the tokenswap period
        // 将众筹的neo转化为等价的ico代币
       [DisplayName("mintTokens")]
        public static bool MintTokens()
        {
            byte[] sender = GetSender();
            // contribute asset is not neo
            if (sender.Length == 0)
            {
                return false;
            }
            ulong contribute_value = GetContributeValue();
            // the current exchange rate between ico tokens and neo during the token swap period
            // 获取众筹期间ico token和neo间的转化率
            ulong swap_rate = CurrentSwapRate();
            // crowdfunding failure
            // 众筹失败
            if (swap_rate == 0)
            {
                Refund(sender, contribute_value);
                return false;
            }
            // you can get current swap token amount
            ulong token = CurrentSwapToken(sender, contribute_value, swap_rate);
            if (token == 0)
            {
                return false;
            }
            // crowdfunding success
            // 众筹成功
            BigInteger balance = Storage.Get(Storage.CurrentContext, sender).AsBigInteger();
            Storage.Put(Storage.CurrentContext, sender, token + balance);
            BigInteger totalSupply = Storage.Get(Storage.CurrentContext, "totalSupply").AsBigInteger();
            Storage.Put(Storage.CurrentContext, "totalSupply", token + totalSupply);
            Transferred(null, sender, token);
            return true;
        }

        // get the total token supply
        // 获取已发行token总量
        [DisplayName("totalSupply")]
        public static BigInteger TotalSupply()
        {
            return Storage.Get(Storage.CurrentContext, "totalSupply").AsBigInteger();
        }

        // function that is always called when someone wants to transfer tokens.
        // 流转token调用
        [DisplayName("transfer")]
        public static bool Transfer(byte[] from, byte[] to, BigInteger value)
        {
            if (value <= 0) return false;
            if (!Runtime.CheckWitness(from)) return false;
            if (to.Length != 20) return false;

            BigInteger from_value = Storage.Get(Storage.CurrentContext, from).AsBigInteger();
            if (from_value < value) return false;
            if (from == to) return true;
            if (from_value == value)
                Storage.Delete(Storage.CurrentContext, from);
            else
                Storage.Put(Storage.CurrentContext, from, from_value - value);
            BigInteger to_value = Storage.Get(Storage.CurrentContext, to).AsBigInteger();
            Storage.Put(Storage.CurrentContext, to, to_value + value);
            Transferred(from, to, value);
            return true;
        }

        // get the account balance of another account with address
        // 根据地址获取token的余额
        [DisplayName("balanceOf")]
        public static BigInteger BalanceOf(byte[] address)
        {
            return Storage.Get(Storage.CurrentContext, address).AsBigInteger();
        }

        // The function CurrentSwapRate() returns the current exchange rate
        // between ico tokens and neo during the token swap period
        private static ulong CurrentSwapRate()
        {
            const int ico_duration = ico_end_time - ico_start_time;
            uint now = Runtime.Time;
            int time = (int)now - ico_start_time;
            if (time < 0)
            {
                return 0;
            }
            else if (time < ico_duration)
            {
                return basic_rate;
            }
            else
            {
                return 0;
            }
        }

        //whether over contribute capacity, you can get the token amount
        private static ulong CurrentSwapToken(byte[] sender, ulong value, ulong swap_rate)
        {
            ulong token = value / neo_decimals * swap_rate;
            BigInteger total_supply = Storage.Get(Storage.CurrentContext, "totalSupply").AsBigInteger();
            BigInteger balance_token = total_amount - total_supply;
            if (balance_token <= 0)
            {
                Refund(sender, value);
                return 0;
            }
            else if (balance_token < token)
            {
                Refund(sender, (token - balance_token) / swap_rate * neo_decimals);
                token = (ulong)balance_token;
            }
            return token;
        }

        // check whether asset is neo and get sender script hash
        private static byte[] GetSender()
        {
            Transaction tx = (Transaction)ExecutionEngine.ScriptContainer;
            TransactionOutput[] reference = tx.GetReferences();
            // you can choice refund or not refund
            foreach (TransactionOutput output in reference)
            {
                if (output.AssetId == neo_asset_id) return output.ScriptHash;
            }
            return new byte[]{};
        }

        // get smart contract script hash
        private static byte[] GetReceiver()
        {
            return ExecutionEngine.ExecutingScriptHash;
        }

        // get all you contribute neo amount
        private static ulong GetContributeValue()
        {
            Transaction tx = (Transaction)ExecutionEngine.ScriptContainer;
            TransactionOutput[] outputs = tx.GetOutputs();
            ulong value = 0;
            // get the total amount of Neo
            // 获取转入智能合约地址的Neo总量
            foreach (TransactionOutput output in outputs)
            {
                if (output.ScriptHash == GetReceiver() && output.AssetId == neo_asset_id)
                {
                    value += (ulong)output.Value;
                }
            }
            return value;
        }
    }
}

File: java/ico-template/src/main/java/org/ico/main/ICOTemplate.java
package org.ico.main;

import org.neo.smartcontract.framework.services.neo.TriggerType;
import org.neo.smartcontract.framework.services.system.ExecutionEngine;

import java.math.BigInteger;

import org.ico.token.NEP5Template;
import org.ico.token.TemplateToken;
import org.neo.smartcontract.framework.Helper;
import org.neo.smartcontract.framework.SmartContract;
import org.neo.smartcontract.framework.services.neo.Runtime;
import org.neo.smartcontract.framework.services.neo.Storage;
import org.neo.smartcontract.framework.services.neo.Transaction;
import org.neo.smartcontract.framework.services.neo.TransactionOutput;

public class ICOTemplate extends SmartContract {

	/**
	 * The EPOC start time for the ICO. The associated time unit is seconds
	 */
	private final static long ICO_START_TIME = 1521003600;

	/**
	 * The EPOC end time for the ICO. The associated time unit is seconds
	 */
	private final static long ICO_END_TIME = 1523682000;

	/**
	 * The Entry point for the ICO Smart Contract. The neoj neo-compiler looks
	 * for a method named: Main with an uppercase M.
	 */
	public static Object Main(String operation, Object[] args) {
		TriggerType trigger = Runtime.trigger();

		if (trigger == TriggerType.Verification) {
			boolean isOwner = Runtime.checkWitness(TemplateToken.getOwner());

			if (isOwner) {
				return true;
			}
			return false;

		} else if (trigger == TriggerType.Application) {

			if (operation.equals("deploy")) {
				return deploy();
			} else if (operation.equals("mintTokens")) {
				return mintTokens();
			} else if (operation.equals(NEP5Template.TOTAL_SUPPLY)) {
				return NEP5Template.totalSupply();
			} else if (operation.equals(NEP5Template.NAME)) {
				return NEP5Template.name();
			} else if (operation.equals(NEP5Template.SYMBOL)) {
				return NEP5Template.symbol();
			} else if (operation.equals(NEP5Template.DECIMALS)) {
				return NEP5Template.decimalsString();
			} else if (operation.equals(NEP5Template.BALANCE_OF)) {
				if (args.length < 1) {
					Runtime.log("No account argument-No action");
					return false;
				}

				byte[] account = Helper.asByteArray((String) args[0]);
				return NEP5Template.balanceOf(account);
			} else if (operation.equals(NEP5Template.TRANSFER)) {
				if (args.length != 3) {
					Runtime.log("Insufficient number of arguments-No action");
					return false;
				}

				byte[] from = Helper.asByteArray((String) args[0]);
				byte[] to = Helper.asByteArray((String) args[1]);
				BigInteger amount = BigInteger.valueOf(Long.valueOf((String) args[2]));

				return NEP5Template.transfer(from, to, amount);
			}
		}

		return false;
	}

	public static Object deploy() {
		if (getTotalSupply().length != 0) {
			Runtime.log("Insufficient token supply-No action");
			return false;
		}

		Storage.put(Storage.currentContext(), TemplateToken.getOwner(), TemplateToken.getPreIcoCap());
		Storage.put(Storage.currentContext(), NEP5Template.TOTAL_SUPPLY, TemplateToken.getPreIcoCap());

		return true;
	}

	public static Object mintTokens() {
		byte[] sender = getSender();

		if (sender.length == 0) {
			Runtime.log("Asset is not neo-No action");
			return false;
		}

		// The current exchange rate between ICO tokens and Neo during the token
		// swap period
		BigInteger swapRate = currentSwapRate();
		if (swapRate.equals(BigInteger.ZERO)) {
			Runtime.log("Crowd funding failure-No action");
			return false;
		}

		BigInteger contributionAmount = getContributionAmount();
		BigInteger tokenTransferAmount = currentSwapToken(sender, contributionAmount, swapRate);
		if (tokenTransferAmount.equals(BigInteger.ZERO)) {
			Runtime.log("Zero token transfer amount-No action");
			return false;
		}

		byte[] senderBalanceByteArray = Storage.get(Storage.currentContext(), sender);
		BigInteger senderBalance = new BigInteger(senderBalanceByteArray);
		BigInteger postTransferAmount = senderBalance.add(tokenTransferAmount);
		Storage.put(Storage.currentContext(), sender, postTransferAmount);

		BigInteger tokenAndSupply = tokenTransferAmount.add(NEP5Template.totalSupply());
		Storage.put(Storage.currentContext(), NEP5Template.TOTAL_SUPPLY, tokenAndSupply);

		return true;
	}

	private static BigInteger currentSwapRate() {
		long icoDuration = ICO_END_TIME - ICO_START_TIME;

		long now = Runtime.time();
		long time = now - ICO_START_TIME;

		if (time < 0) {
			return BigInteger.ZERO;
		} else if (time < icoDuration) {
			return TemplateToken.getBasicRate();
		}

		return BigInteger.ZERO;
	}

	private static BigInteger getContributionAmount() {
		Transaction tx = (Transaction) ExecutionEngine.scriptContainer();
		TransactionOutput[] outputs = tx.outputs();
		BigInteger contributionAmount = BigInteger.ZERO;

		for (TransactionOutput transaction : outputs) {
			if (transaction.scriptHash() == getReceiver() && transaction.assetId() == neoAssetId()) {
				BigInteger transactionAmount = BigInteger.valueOf(transaction.value());
				contributionAmount = contributionAmount.add(transactionAmount);
			}
		}

		return contributionAmount;
	}

	private static BigInteger currentSwapToken(byte[] sender, BigInteger contributedAmount, BigInteger swapRate) {
		BigInteger tokenAmount = contributedAmount.divide(TemplateToken.getDecimals()).multiply(swapRate);
		BigInteger tokenBalance = TemplateToken.getTotalAmount().subtract(NEP5Template.totalSupply());

		if (tokenBalance.compareTo(BigInteger.ZERO) <= 0) {
			return BigInteger.ZERO;
		} else if (tokenBalance.compareTo(tokenAmount) < 0) {
			tokenAmount = tokenBalance;
		}

		return tokenAmount;
	}

	/**
	 * @return smart contract script hash
	 */
	private static byte[] getReceiver() {
		return ExecutionEngine.executingScriptHash();
	}

	private static byte[] getSender() {
		Transaction tx = (Transaction) ExecutionEngine.scriptContainer();
		TransactionOutput[] reference = tx.references();

		for (TransactionOutput output : reference) {
			if (output.assetId() == neoAssetId()) {
				return output.scriptHash();
			}
		}
		return Helper.asByteArray("");
	}

	private static byte[] getTotalSupply() {
		return Storage.get(Storage.currentContext(), NEP5Template.TOTAL_SUPPLY);
	}

	private static byte[] neoAssetId() {
		byte[] assetId = new byte[] { (byte) 155, (byte) 124, (byte) 255, (byte) 218, (byte) 166, (byte) 116,
				(byte) 190, (byte) 174, (byte) 15, (byte) 147, (byte) 14, (byte) 190, (byte) 96, (byte) 133, (byte) 175,
				(byte) 144, (byte) 147, (byte) 229, (byte) 254, (byte) 86, (byte) 179, (byte) 74, (byte) 92, (byte) 34,
				(byte) 12, (byte) 205, (byte) 207, (byte) 110, (byte) 252, (byte) 51, (byte) 111, (byte) 197 };
		return assetId;
	}

}
